Performance optimizations #1

frankmayer opened the pull request:
**[UPDATED (5.1.2017) - see second paragraph]**
Those are some performance optimizations for the libraries/cms folder. If you like those, there are plenty more where those came from.
Since I don't know if you guys are open to these kinds of PR's. I am awaiting the developments of my open PR's before I do more.

**[Update 5.1.2017]**
In order to lighten this PR up, I have introduced 6 sub-PRs, which are easier to digest and mostly with very specific changes. In order to continue with this one, those ones should be tested/reviewed/merged first. After that, I can resolve the conflicts and there should finally be only a few changes left.

- [x] 1) Simplify some ternary operations using elvis operator and remove unnecessary parentheses in libraries/cms #13234
- [x] 2) Replace unnecessary double quotes in libraries/cms #13235
- [x] 3) Remove one-time-use variables in libraries/cms #13236
- [x] 4) Various changes in libraries/cms #13238
- [x] 5) Type safe comparison of strings only, in libraries/cms #13275
- [x] 6) Type safe comparison in libraries/cms - second iteration #13276

frankmayer committed:
Fixed callable call in loop termination condition

frankmayer committed:
Fixed non-optimal regular expression

frankmayer committed:
Shortened syntax for applied operations

frankmayer committed:
Don't use strlen() to check if string is empty.

zero-24 reviewed (approved):
looks good from here. Thanks!

zero-24 reviewed (changes requested):
Please have a look at: https://travis-ci.org/joomla/joomla-cms/jobs/162592381 It looks like travis is not happy.

zero-24 commented:
Travis error is:

```
1) JComponentRouterLegacyTest::testParse
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     'option' => 'com_test'
-    42 => 'test-test'
+    42 => 'test:test'
 )
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/cms/component/router/JComponentRouterLegacyTest.php:130
2) JRouterTest::testDecodeSegments with data set #3 (array('42-test', 'testing-this-method'), array('42:test', 'testing:this-method'))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     0 => '42:test'
-    1 => 'testing:this-method'
+    1 => 'testing:this:method'
 )
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/cms/router/JRouterTest.php:872
```

https://travis-ci.org/joomla/joomla-cms/jobs/162592381

frankmayer committed:
Reverted preg changes. Look into those later again.

frankmayer commented:
Yes, I noticed... :) Will fix errors.

frankmayer committed:
Replaced call_user_func_array() with call_user_func()

frankmayer committed:
Merge unset() calls

frankmayer committed:
Replaced old intval() call with modern typecasting

frankmayer committed:
Optimize away substr() as index-based access

frankmayer committed:
Merge branch 'Performance_2' into Performance_1

frankmayer committed:
Replace some cases of substr() with strpos()

frankmayer committed:
Replace some stristr() with stripos()

frankmayer committed:
Replace is_null() with null === ...

frankmayer committed:
Removed unnecessary ternary operators

frankmayer commented:
on it...

frankmayer committed:
Fix for Juri...

frankmayer committed:
Revert some changes for ... PHP5.3 compatibility

frankmayer committed:
Make travis happy ....

frankmayer committed:
Make codesniffer happy...

frankmayer committed:
DYC!

frankmayer committed:
Optimize non-optimal if conditions.

frankmayer committed:
codsniffer...

frankmayer committed:
Inline one-time use variables

frankmayer committed:
Codesniffer...

frankmayer commented:
That about sums this PR up. No new commits, as far as optimizations are concerned.
Let me know if you need me to change anything.

photodude commented on libraries/cms/application/site.php:
```diff
@@ -269,9 +269,7 @@ public function getLanguageFilter()
-		$menu = parent::getMenu($name, $options);
-
-		return $menu;
+		return parent::getMenu($name, $options);
```
I can't find the reference at the moment, but I remember reading something along the lines that storing a function return in a variable and returning that variable in your method is less error prone and faster than directly returning an evaluated function in your method. 

Technically they both work, and off the top of my head I don't think we have a standard on which way to write code.

frankmayer replied:
In my opinion, such kind of variables should not be introduced. 

They serve no practical purpose. And they do actually introduce another unnecessary step in our minds, when we humans read the code - and also when the machine aka the interpreter/compiler reads and processes it. 

Why introduce something whose only purpose is to be returned on the next line, without any further processing?

As for the compiler, I would think modern PHP compilers do this step automatically internally, but they have to do it nevertheless, using cpu and memory, so why not leave it out in the first place?

Just my 2 cents :smile:

mbabker replied:
In terms of code readability I've been leaning toward less creation/assignment of variables that aren't needed than "explicit" structures like this one.  There isn't a standard since both are valid syntax, but unless there is a majorly strong argument (like engine performance) for one over the other, I'd lean toward the readability argument and I find it easier to read one line instead of two code lines and the blank line in the middle.

frankmayer replied:
BTW, there are some interesting numbers [here](http://obrown.io/php/2015/12/19/php-performance-inlining.html) on inlining and redundant variables. These numbers depict different implementations in loops, but you can see the difference, which is even bigger in 7.0 (than even the latest 5.6), when using redundant variables.
Of course, it's not like Joomla is looping several thousands of times through these functions to create some content, but why not save a few bytes on memory and also a few cpu cycles? We can only benefit from it. ... + it's easier to read...

frankmayer replied:
@photodude I think you were possibly referring to this q&a...

photodude replied:
That might have been it, but maybe not. I just checked some of my book references to see if it was in one of those books, and it seems the general consensus in the examples leans towards the inlining the return.

The argument of code readability is a good one. 

It might be one of those specific circumstances things. Like ternary operators are ok in php, unless there is large objects in the variables being passed, then ternary operators in php suck up time and memory due to the copy operation they preform on the variables before evaluating. 

In anycase I can't seem to find the reference at the moment, so Code Readability wins me over.

photodude replied:
@frankmayer I think part of what I remember was something like this comment on the PHP manual
http://php.net/manual/en/function.pg-connect.php#82117
considering that it's 8 years old I'm not sure if it still applies in more modern php

frankmayer replied:
@photodude I see what you mean. The case of the comment is partially true (at least with PHP7 which I just tested). In PHP7, it will die() even inline. However, if the returned value is not false, it will return a bool, which automatically will be `true`.
So this is actually a case where caution is advised and those cases should be dealt with, the old way.

sovainfo commented:
Would you mind motivating the other types of performance improvement as well?

frankmayer commented:
@sovainfo I'll gladly point out my motivation for this. 

Generally, all of the improvements are based on modern PHP best practices.

Most of them are dropping function calls in favor of using cheaper (in terms of cpu usage and memory) implementations.

Often, readability is also improved as a welcome side-effect of this.

Other things like the optimized if-conditions, are moving the cheaper conditions to the beginning of the conditions, in order to not invoke more expensive operations, if not necessary.

Of course there are some situations, where optimization is not possible, because of the necessity of maintaining PHP 5.3 compatibility. (Are there really people still running 5.3???)
IMHO, Joomla 4, whenever it comes out, should radically drop the complete 5.* line of PHP and have a minimum of PHP 7.0. Because 5.6 is going into 'security fixes only' mode on 1.1.2017 - [see here](http://php.net/supported-versions.php)

That should about sum it up.

As you have probably noticed, this PR is only focusing on the files in libraries/cms + it's subfolders, as I though it was a good starting point. If you guys are OK with those changes, I'll gladly expand the implementation of that kind of optimizations and cleanup to the whole project.

sovainfo commented:
Thanks, was wondering about the optimization of if-conditions. 

The costs of the condition shouldn't be the only factor. Sounds like you want to ignore the likely hood of the conditions, together they should determine the order. 
This assumes it was given thought in the first place.

frankmayer commented:
@sovainfo the cost actually isn't the only factor. For that reason, not all if-conditions were touched. I hand-picked only those, were it would not change their semantics. Did you notice some inconsistency after those changes?

photodude commented:
Surprisingly almost 14% of submitted user statistics were for PHP 5.3
I personally question how many of those were submitted by developers doing testing
https://developer.joomla.org/about/stats.html

frankmayer commented:
@photodude yes, that's a valid question. I too think the developer tests are tinkering with real statistics. I wonder if there should be some kind of 'isDeveloper' switch, which developers could set, in order to not have unwanted side effects like this.

Bakual commented:
Personally I think almost no dev tests with 5.3. Devs are the ones with the PHP7 servers :smile: 
But users really **are** still using PHP 5.3. If you ask why, it's because they don't care about underlying systems and just trust their hosters. And thos don't want to risk breaking sites by upgrading the PHP version.

mbabker commented:
Part of it is we didn't want to make a distinction between dev and production environments in the statistics.  It honestly doesn't matter to us either (well, that's my opinion anyway, YMMV).  That's in part why you get the nag message versus default enabled (every fresh install I do I immediately hit never on it).  We have provisions in the stats collection making it possible to report things that nobody will be using in production (i.e. our 4.0 branch doesn't get rejected, if you're compiling against PHP master which has a 7.2 version number right now it's accepted, etc.) so that too can skew things because it favors either ambitious or curious folks who want to try things out on bleeding edge technology and that can also be used as an indicator to see what folks are doing.

frankmayer commented:
@Bakual I don't know about the devs testing with 5.3, but if the number really depicts users still using 5.3, then that's one more reason to require PHP 7.0 for Joomla 4, because if Joomla 4 would be released in 2017, then PHP 5.6 would already be in security only-fixes mode for about a year until its final EOL end of 2018.

This is Joomla's chance to help hosters get rid of old versions and also be able to use the newer technologies and speed provided by the PHP 7 series.

Bakual commented:
I don't think it's reasonable to raise the minimum PHP requirement for Joomla 4 to PHP 7, but PHP 5,.6 for sure. If it is released next year that is.
By the same logic you should raise the minimum requirement to PHP 7.1, because security support for PHP 7.0 actually runs out before PHP 5.6 does :smile: 

But it is not up to me to make that decision.

photodude commented:
@frankmayer In the past (and currently) there is a dynamic between what hosters are supporting (influenced by what's available and supported in standard distros) vs what developers want. Personally, I would love to see J3.7 drop php5.3 but that likely won't happen. On the other hand J4 is being developed with a php5.6+ design. Sadly most hosters are still hesitant to fully support php 7 even when they offer the option to use php7 (when problems arise they typically tell people to go back to using php5.6, even some developers of extensions tell customers that)

Balancing the hoster support, 3rd party extensions, and platform needs is an unpleasant challenge. IMO php5.6 is a good compromize for J4.

matrikular commented on libraries/cms/application/cms.php:
```diff
@@ -114,13 +114,13 @@ public function __construct(JInput $input = null, Registry $config = null, JAppl
 
 		// Enable sessions by default.
-		if (is_null($this->config->get('session')))
+		if (null === $this->config->get('session'))
```
If there there is no reason other than taste, I would propose to flip the expression condition to, in this case; `if ($this->config->get('session') === null)` or, for line 584; `if ($registry !== null)`

frankmayer replied:
In the past I was writing it that way, but recently I discovered I find it easier to read it this way.
It reads like `null is something`, which personally, I find easier to read and process than `something is null`.
However, if the consensus is against that style, I'll happily flip those cases.

Bakual replied:
I don't think it is explicitely set in the coding guidelines but it sure is the way we usually write it.  So if you don't mind, please flip it.

frankmayer replied:
Ok, will do.

frankmayer replied:
Done.

photodude replied:
Yoda conditionals always look odd to me too.  Maybe it's all of the Left-to-right reading/writing in English that influences that oddness.

matrikular commented on libraries/cms/menu/menu.php:
```diff
@@ -229,9 +229,8 @@ public function setActive($id)
-			$result = &$this->_items[$id];
 
-			return $result;
+			return $this->_items[$id];
```
@frankmayer was it a conscious decision to "remove" the reference at this point? Maybe someone can shed some light on the old implementation, too.

frankmayer replied:
Yes if I remember well, it was.
I just checked again and I must have figured that in this case an object would be returned, so I removed the reference, since objects are returned by reference anyways (or in other words, a pointer to the object is returned). 

I remember also seeing other remnants from the PHP4 days like this, but I din't touch these, unless they were part of a wanted change, like the one in question.

matrikular replied:
Thank you for the clarification.

frankmayer committed:
Flipped comparisons
`null === ...`  to Joomla Style `... === null`
`null !== ...`  to Joomla Style `... !== null`

matrikular commented on libraries/cms/helper/media.php:
```diff
@@ -206,7 +206,7 @@ public function canUpload($file, $component = 'com_media')
 		{
 			// A tag is '<tagname ', so we need to add < and a space or '<tagname>'
-			if (stristr($xss_check, '<' . $tag . ' ') || stristr($xss_check, '<' . $tag . '>'))
+			if (false !== stripos($xss_check, '<' . $tag . ' ') || false !== stripos($xss_check, '<' . $tag . '>'))
```
@frankmayer regarding the flipping (my ocd) this might be changed to ...
`if (stripos($xss_check, '<' . $tag . ' ') !== false [...]`
... as well, if you may.

frankmayer replied:
@matrikular whoops! sorry, those ones slipped through... will do

matrikular commented on libraries/cms/schema/changeitem/postgresql.php:
```diff
@@ -86,7 +86,7 @@ protected function buildCheckQuery()
 					$type = '';
 
-					for ($i = 7; $i < count($wordArray); $i++)
+					for ($i = 7, $iMax = count($wordArray); $i < $iMax; $i++)
```
@frankmayer good find.

matrikular commented:
From what I could grasp by reviewing the code alone, all looks good.

There are some areas, like database queries without a try / catch block neither before or after the optimisation, but I guess that alone might be a thing for different PR. Anonther thing that caught my attention is [this passage](https://github.com/frankmayer/joomla-cms/blob/062ef22f7608f82d7fe47cfe9f6b20ea13d0e1e1/libraries/cms/application/site.php#L607-L660 "JApplicationSite::initialiseApp") from JApplicationSite::initialiseApp().

Good work, thanks!

frankmayer committed:
Flipped comparisons
`false !== ...`  to Joomla Style `... !== falsel`

frankmayer commented:
@matrikular You're very welcome! I always wanted to give back to the Joomla project, and decided to give it a try by doing this. And as I mentioned earlier, if you (as the steering users) are OK with those kinds of changes I am open to expand the scope of optimizations to the whole project (minus the vendor imports... as I learned :smile: )

chrisdavenport commented on libraries/cms/html/tabs.php:
```diff
@@ -86,7 +87,7 @@ protected static function loadBehavior($group, $params = array())
 
 			// When use storage is set and value is false - By default we allow to use storage
-			$opt['useStorage'] = (isset($params['useCookie']) && !$params['useCookie']) ? false : true;
+			$opt['useStorage'] = (isset($params['useCookie']) && !$params['useCookie']);
```
This doesn't look right.  I think you need to do
`$opt['useStorage'] = !(isset($params['useCookie']) && !$params['useCookie']);`

frankmayer replied:
Good catch!! Yes. my bad. will fix. Thanks

chrisdavenport commented on libraries/cms/installer/adapter/component.php:
```diff
@@ -356,7 +356,7 @@ public function getElement($element = null)
 		$element = parent::getElement($element);
 
-		if (substr($element, 0, 4) != 'com_')
+		if (0 !== strpos($element, 'com_'))
```
I can see what you're getting at, but it looks really ugly (to me).  Is the performance gain really worth it?

frankmayer replied:
That's a big discussion, but IMHO why not do it? It doesn't cost anything.
On the contrary, any gain no matter how big or small will add to the pile and help reduce load on the machine it's running on. What do you think?

chrisdavenport commented on libraries/cms/module/helper.php:
```diff
@@ -50,7 +50,7 @@ public static function &getModule($name, $title = null)
 
 		// If we didn't find it, and the name is mod_something, create a dummy object
-		if (is_null($result) && substr($name, 0, 4) == 'mod_')
+		if ($result === null && 0 === strpos($name, 'mod_'))
```
Same here.  Maybe `strpos($name, 'mod_') === 0` will make it look prettier. ;-)

frankmayer replied:
Yes it will ;)

frankmayer committed:
Fix removal of ternary operator. Forgot inversion.

frankmayer committed:
More flipping of  comparisons ;)

frankmayer committed:
Removed another unnecessary set of ternary operators.

frankmayer committed:
Codesniffer...

andrepereiradasilva commented on libraries/cms/helper/route.php:
```diff
@@ -180,7 +180,7 @@ protected function findItem($needles = array())
 						 * $language == * cannot override existing entries
 						 */
-						if (!isset(static::$lookup[$language][$view][$item->query['id']]) || $item->language != '*')
+						if ($item->language != '*' || !isset(static::$lookup[$language][$view][$item->query['id']]))
```
missed one !== here

frankmayer replied:
Missed a lot actually... not sure if I was so eager to include that many changes in that first PR.

andrepereiradasilva commented on libraries/cms/plugin/plugin.php:
```diff
@@ -110,7 +110,7 @@ public function __construct(&$subject, $config = array())
 			$dbProperty = $reflection->getProperty('db');
 
-			if ($dbProperty->isPrivate() === false && is_null($this->db))
+			if ($dbProperty->isPrivate() === false && $this->db === null)
```
not sure of this, but couldn't we do it like this
```php
if ($this->db === null && $reflection->getProperty('db')->isPrivate() === false)
```

andrepereiradasilva commented on libraries/cms/plugin/plugin.php:
```diff
@@ -99,7 +99,7 @@ public function __construct(&$subject, $config = array())
 			$appProperty = $reflection->getProperty('app');
 
-			if ($appProperty->isPrivate() === false && is_null($this->app))
+			if ($appProperty->isPrivate() === false && $this->app === null)
```
not sure of this, but couldn't we do it like this
```php
if ($this->app === null && $reflection->getProperty('app')->isPrivate() === false)
```

frankmayer replied:
Should be no problem.

andrepereiradasilva commented on libraries/cms/search/helper.php:
```diff
@@ -43,7 +43,7 @@ public static function logSearch($term, $component)
 				->where($db->quoteName('search_term') . ' = ' . $db->quote($search_term));
 			$db->setQuery($query);
-			$hits = intval($db->loadResult());
+			$hits = (int) ($db->loadResult());
```
can't we remove the parenthisis here?

frankmayer committed:
Type-safe comparisons on strings

frankmayer committed:
More type-safe comparisons

frankmayer committed:
Changes that occurred during PR discussion

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on 171f66b0ee535c1103a5e7ee94b58ed8d8082e60<br /><br />on code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12171">issues.joomla.org/tracker/joomla-cms/12171</a>.</sub>

frankmayer committed:
Merge branch 'staging' into Performance_1

# Conflicts:
#	libraries/cms/component/helper.php
#	libraries/cms/html/behavior.php
#	libraries/cms/html/html.php
#	libraries/cms/router/site.php

frankmayer committed:
Remove some unnecessary parentheses

frankmayer committed:
Removed empty if-group and reversed condition

frankmayer commented:
Fixed some conflicts that had occurred in the meantime. It's essentially ready to be merged, when the team finds the time and PR is OK.

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on 171f66b0ee535c1103a5e7ee94b58ed8d8082e60<br /><br />code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12171">issues.joomla.org/tracker/joomla-cms/12171</a>.</sub>

frankmayer committed:
Merge branch 'staging' into Performance_1

andrepereiradasilva commented:
conflicts

frankmayer commented:
@andrepereiradasilva Yes, but after the referenced open PR's are closed. Don't want to do this three times ;)

andrepereiradasilva commented:
right i see

frankmayer commented:
Edited initial post to show which sub-PRs still have to go through, in order to make this one easier to digest. 
It would be great if the initial big batch PRs of mine could make it into alpha2. Thanks!

frankmayer committed:
Merge branch 'staging' into Performance_1

# Conflicts:
#	libraries/cms/editor/editor.php
#	libraries/cms/form/field/limitbox.php
#	libraries/cms/form/field/menu.php
#	libraries/cms/helper/media.php
#	libraries/cms/html/access.php
#	libraries/cms/html/jgrid.php
#	libraries/cms/html/menu.php
#	libraries/cms/html/string.php
#	libraries/cms/installer/adapter/package.php
#	libraries/cms/installer/installer.php
#	libraries/cms/router/site.php
#	libraries/cms/schema/changeitem.php
#	libraries/cms/schema/changeitem/mysql.php
#	libraries/cms/schema/changeset.php

frankmayer committed:
Merge branch 'staging' into Performance_1

# Conflicts:
#	libraries/cms/application/administrator.php
#	libraries/cms/application/cms.php
#	libraries/cms/application/helper.php
#	libraries/cms/application/site.php
#	libraries/cms/form/field/editor.php
#	libraries/cms/form/field/limitbox.php
#	libraries/cms/form/field/tag.php
#	libraries/cms/helper/media.php
#	libraries/cms/helper/route.php
#	libraries/cms/html/grid.php
#	libraries/cms/html/html.php
#	libraries/cms/html/searchtools.php
#	libraries/cms/html/select.php
#	libraries/cms/html/string.php
#	libraries/cms/installer/adapter/component.php
#	libraries/cms/installer/adapter/plugin.php
#	libraries/cms/installer/installer.php
#	libraries/cms/installer/script.php
#	libraries/cms/module/helper.php
#	libraries/cms/plugin/helper.php
#	libraries/cms/schema/changeitem/mysql.php
#	libraries/cms/schema/changeitem/postgresql.php
#	libraries/cms/schema/changeitem/sqlsrv.php
#	libraries/cms/table/corecontent.php
#	libraries/cms/toolbar/button/popup.php

frankmayer committed:
Corrections on errors during merge (Conflict resolving)

frankmayer committed:
Missed one.

frankmayer committed:
Remove a parenthesis

andrepereiradasilva commented on libraries/cms/helper/media.php:
```diff
@@ -360,15 +360,15 @@ public function countFiles($dir)
 			{
-				if (substr($entry, 0, 1) !== '.' && is_file($dir . DIRECTORY_SEPARATOR . $entry)
+				if ($entry[0] !== '.' && is_file($dir . DIRECTORY_SEPARATOR . $entry)
 					&& strpos($entry, '.html') === false && strpos($entry, '.php') === false)
```
IMO `is_file` should be the last check sine it makes I/O call to the filesystem.

frankmayer replied:
Yes, missed that one. Nice catch.

andrepereiradasilva commented on libraries/cms/html/tabs.php:
```diff
@@ -9,6 +9,7 @@
 
 defined('JPATH_PLATFORM') or die;
 
+
```
please remvoe this extra line

andrepereiradasilva commented on libraries/cms/installer/installer.php:
```diff
@@ -2223,7 +2223,8 @@ public static function parseXMLInstallFile($path)
 
-		if ($xml->getName() !== 'extension' && $xml->getName() !== 'metafile')
+		$name = $xml->getName();
+		if ($name !== 'extension' && $name !== 'metafile')
```
cs: add a return before the if statement

andrepereiradasilva commented on libraries/cms/plugin/plugin.php:
```diff
@@ -96,9 +96,8 @@ public function __construct(&$subject, $config = array())
-			$appProperty = $reflection->getProperty('app');
 
-			if ($appProperty->isPrivate() === false && is_null($this->app))
+			if ($reflection->getProperty('app')->isPrivate() === false && $this->app === null)
```
change the order of the contitions?

frankmayer replied:
Yep

andrepereiradasilva commented on libraries/cms/plugin/plugin.php:
```diff
@@ -107,9 +106,8 @@ public function __construct(&$subject, $config = array())
-			$dbProperty = $reflection->getProperty('db');
 
-			if ($dbProperty->isPrivate() === false && is_null($this->db))
+			if ($reflection->getProperty('db')->isPrivate() === false && $this->db === null)
```
change the order of the contitions?

andrepereiradasilva reviewed (commented):
some minor comments

frankmayer committed:
Changes according to reviewer's comment

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on 171f66b0ee535c1103a5e7ee94b58ed8d8082e60<br /><br />code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12171">issues.joomla.org/tracker/joomla-cms/12171</a>.</sub>

peteruoi commented:
Just an idea...
If there is a considerable performance in joomla paging loading maybe a metric could be done with and without all these performance prs and if is enough it could be advertises in joomla 3.8 release campaign?
thanks frankmayer for your work!

frankmayer committed:
Revert change because of failure

frankmayer commented:
@peteruoi that is an interesting idea. However I assume that any improvements from all the PRs will vary a lot with mileage. Unfortunately I haven't got the time to set this up and do the tests.
Oh, and a lot of things made it already into 3.7.

Another situation  where I would hope to see some measurable optimization would be, if the indexer performance PR would be merged (see https://github.com/joomla/joomla-cms/pull/13511).
This would of course mostly be measureable with initial (or re-) indexing of a lot of index-able content. 
But I also would like to think that on changing of content, it might work a little bit faster.

frankmayer commented on libraries/cms/html/menu.php:
```diff
@@ -192,7 +192,7 @@ public static function menuItemList($name, $selected = null, $attribs = null, $c
 			array(
 				'id' => isset($config['id']) ? $config['id'] : 'assetgroups_' . (++$count),
-				'list.attr' => (is_null($attribs) ? 'class="inputbox" size="1"' : $attribs),
+				'list.attr' => ($attribs === null ? 'class="inputbox" size="1"' : $attribs),
```
Yes, that's just the right place ;)

frankmayer replied:
Got more? Shall I wait?

Quy replied:
Also align `=`

Quy replied:
More reported and I am done for now. :wink:

frankmayer replied:
Thanks. Done a few more things I found along the way.

Quy commented on libraries/cms/pagination/pagination.php:
```diff
@@ -451,7 +451,7 @@ public function getPagesLinks()
 	{
 		// Allow to receive a null layout
-		$layoutId = (null === $layoutId) ? 'joomla.pagination.links' : $layoutId;
+		$layoutId = ($layoutId === null) ? 'joomla.pagination.links' : $layoutId;
```
Remove parentheses and more below.

Quy commented on libraries/cms/table/contenthistory.php:
```diff
@@ -127,12 +127,12 @@ public function getSha1($jsonData, JTableContenttype $typeTable)
 				foreach ($value as $subName => $subValue)
 				{
-					$object->$subName = (is_int($subValue) || is_bool($subValue) || is_null($subValue)) ? (string) $subValue : $subValue;
+					$object->$subName = (is_int($subValue) || is_bool($subValue) || $subValue === null) ? (string) $subValue : $subValue;
```
Remove parentheses

Quy commented on libraries/cms/table/contenthistory.php:
```diff
@@ -127,12 +127,12 @@ public function getSha1($jsonData, JTableContenttype $typeTable)
 			else
 			{
-				$object->$name = (is_int($value) || is_bool($value) || is_null($value)) ? (string) $value : $value;
+				$object->$name = (is_int($value) || is_bool($value) || $value === null) ? (string) $value : $value;
```
Remove parentheses

frankmayer committed:
Changes according to reviewer's comment
+ a few more changes found last minute

frankmayer committed:
reversed change, though according to the method's doc that value should have been an integer...

frankmayer committed:
Hopefully fixed...

frankmayer committed:
Hopefully fixed...

frankmayer committed:
Re-apply some changes

frankmayer commented:
Bump @andrepereiradasilva & @Quy Problem solved. Pls test (review changes) to go for RTC

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on 171f66b0ee535c1103a5e7ee94b58ed8d8082e60<br /><br />Code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12171">issues.joomla.org/tracker/joomla-cms/12171</a>.</sub>

andrepereiradasilva commented:
@peteruoi when we have levels like Joomla server side performance things are not like 30% Faster in all Pages. Is really relative to each change and to the amount of scenarios that change apllies to
For instance in the past made several changes that boosted performance only in multilingual sites.

Imo the most important performance improvements are the ones done to the critical page that apllies to all pages and also most used parts (com_content, jroute, etc). 
That said general improvements like this set of prs also contribute as a whole.

Quy commented:
I have tested this item :white_check_mark: successfully on 171f66b0ee535c1103a5e7ee94b58ed8d8082e60<br /><br />Code review.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12171">issues.joomla.org/tracker/joomla-cms/12171</a>.</sub>

franz-wohlkoenig commented:
RTC after two successful tests.

peteruoi commented:
@andrepereiradasilva Even with a 5% percent general it will be awesome news for advertisment for 3.8 campaign. And no one forbids us to say 5% to 20% (at miltilingual for example).
I think it's a pity that the optimization you awesome fellows do and make joomla and the internet faster is not recognized enough :). 
This metric of course it cannot be pr dependent but before release 3.7 vs 3.8 tests and if they are favourite for us we advertise ;)

frankmayer commented:
Hooray, this really big chunk was finally merged! Thank you all for your comments and efforts in testing
